File: src/main/java/main/tut15/ManyImages.java

package main.tut15;

import com.jogamp.newt.event.KeyEvent;
import com.jogamp.opengl.GL3;
import com.jogamp.opengl.util.GLBuffers;
import com.jogamp.opengl.util.texture.spi.DDSImage;
import glm.mat.Mat4;
import glm.vec._3.Vec3;
import main.framework.Framework;
import main.framework.Semantic;
import main.framework.component.Mesh;
import main.tut14.PerspectiveInterpolation;
import org.xml.sax.SAXException;
import uno.glm.MatrixStack;
import uno.time.Timer;

import javax.xml.parsers.ParserConfigurationException;
import java.io.File;
import java.io.IOException;
import java.net.URISyntaxException;
import java.nio.ByteBuffer;
import java.nio.FloatBuffer;
import java.nio.IntBuffer;
import java.util.logging.Level;
import java.util.logging.Logger;

import static com.jogamp.opengl.GL.GL_BACK;
import static com.jogamp.opengl.GL.GL_BGRA;
import static com.jogamp.opengl.GL.GL_CULL_FACE;
import static com.jogamp.opengl.GL.GL_CW;
import static com.jogamp.opengl.GL.GL_DEPTH_TEST;
import static com.jogamp.opengl.GL.GL_DYNAMIC_DRAW;
import static com.jogamp.opengl.GL.GL_LEQUAL;
import static com.jogamp.opengl.GL.GL_LINEAR;
import static com.jogamp.opengl.GL.GL_LINEAR_MIPMAP_LINEAR;
import static com.jogamp.opengl.GL.GL_LINEAR_MIPMAP_NEAREST;
import static com.jogamp.opengl.GL.GL_MAX_TEXTURE_MAX_ANISOTROPY_EXT;
import static com.jogamp.opengl.GL.GL_NEAREST;
import static com.jogamp.opengl.GL.GL_REPEAT;
import static com.jogamp.opengl.GL.GL_RGB;
import static com.jogamp.opengl.GL.GL_RGB8;
import static com.jogamp.opengl.GL.GL_TEXTURE0;
import static com.jogamp.opengl.GL.GL_TEXTURE_2D;
import static com.jogamp.opengl.GL.GL_TEXTURE_MAG_FILTER;
import static com.jogamp.opengl.GL.GL_TEXTURE_MAX_ANISOTROPY_EXT;
import static com.jogamp.opengl.GL.GL_TEXTURE_MIN_FILTER;
import static com.jogamp.opengl.GL.GL_TEXTURE_WRAP_S;
import static com.jogamp.opengl.GL.GL_TEXTURE_WRAP_T;
import static com.jogamp.opengl.GL.GL_UNPACK_ALIGNMENT;
import static com.jogamp.opengl.GL.GL_UNSIGNED_BYTE;
import static com.jogamp.opengl.GL2ES3.*;
import static com.jogamp.opengl.GL2GL3.GL_UNSIGNED_INT_8_8_8_8_REV;
import static com.jogamp.opengl.GL3.GL_DEPTH_CLAMP;
import static glm.GlmKt.glm;
import static uno.buffer.UtilKt.destroyBuffer;
import static uno.buffer.UtilKt.destroyBuffers;
import static uno.glsl.UtilKt.programOf;


/**
 * @author gbarbieri
 */

public class ManyImages extends Framework {

    public static void main(String[] args) {
        new ManyImages().setup("Tutorial 15 - Many Images");
    }

    private ProgramData program;

    private Mesh plane, corridor;

    private IntBuffer projBufferName = GLBuffers.newDirectIntBuffer(1);

    interface Texture {
        int Checker = 0;
        int MipmapTest = 1;
        int MAX = 2;
    }

    private IntBuffer textureName = GLBuffers.newDirectIntBuffer(Texture.MAX);
    private IntBuffer samplerName = GLBuffers.newDirectIntBuffer(Sampler.MAX);

    private Timer camTimer = new Timer(Timer.Type.Loop, 5f);
    private boolean useMipmapTexture = false;
    private int currSampler = 0;
    private boolean drawCorridor = false;

    @Override
    public void init(GL3 gl) {

        initializeProgram(gl);

        try {
            plane = new Mesh(gl, getClass(), "tut15/BigPlane.xml");
            corridor = new Mesh(gl, getClass(), "tut15/Corridor.xml");
        } catch (ParserConfigurationException | SAXException | IOException | URISyntaxException ex) {
            Logger.getLogger(PerspectiveInterpolation.class.getName()).log(Level.SEVERE, null, ex);
        }

        gl.glEnable(GL_CULL_FACE);
        gl.glCullFace(GL_BACK);
        gl.glFrontFace(GL_CW);

        float depthZNear = 0f;
        float depthZFar = 1f;

        gl.glEnable(GL_DEPTH_TEST);
        gl.glDepthMask(true);
        gl.glDepthFunc(GL_LEQUAL);
        gl.glDepthRangef(depthZNear, depthZFar);
        gl.glEnable(GL_DEPTH_CLAMP);

        //Setup our Uniform Buffers
        gl.glGenBuffers(1, projBufferName);
        gl.glBindBuffer(GL_UNIFORM_BUFFER, projBufferName.get(0));
        gl.glBufferData(GL_UNIFORM_BUFFER, Mat4.SIZE, null, GL_DYNAMIC_DRAW);

        gl.glBindBufferRange(GL_UNIFORM_BUFFER, Semantic.Uniform.PROJECTION, projBufferName.get(0), 0, Mat4.SIZE);

        gl.glBindBuffer(GL_UNIFORM_BUFFER, 0);

        // Generate all the texture names
        gl.glGenTextures(Texture.MAX, textureName);

        loadCheckerTexture(gl);
        loadMipmapTexture(gl);
        createSamplers(gl);
    }

    private void initializeProgram(GL3 gl) {
        program = new ProgramData(gl, "pt.vert", "tex.frag");
    }

    private void loadCheckerTexture(GL3 gl) {

        try {
            File file = new File(getClass().getResource("/tut15/checker.dds").toURI());

            DDSImage ddsImage = DDSImage.read(file);

            gl.glBindTexture(GL_TEXTURE_2D, textureName.get(Texture.Checker));

            for (int mipmapLevel = 0; mipmapLevel < ddsImage.getNumMipMaps(); mipmapLevel++) {

                DDSImage.ImageInfo mipmap = ddsImage.getMipMap(mipmapLevel);

                gl.glTexImage2D(GL_TEXTURE_2D, mipmapLevel, GL_RGB8, mipmap.getWidth(), mipmap.getHeight(), 0,
                        GL_BGRA, GL_UNSIGNED_INT_8_8_8_8_REV, mipmap.getData());
            }

            gl.glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_BASE_LEVEL, 0);
            gl.glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAX_LEVEL, ddsImage.getNumMipMaps() - 1);
            gl.glBindTexture(GL_TEXTURE_2D, 0);

        } catch (IOException | URISyntaxException ex) {
            Logger.getLogger(ManyImages.class.getName()).log(Level.SEVERE, null, ex);
        }
    }

    private void loadMipmapTexture(GL3 gl) {

        gl.glBindTexture(GL_TEXTURE_2D, textureName.get(Texture.MipmapTest));

        IntBuffer oldAlign = GLBuffers.newDirectIntBuffer(1);

        gl.glGetIntegerv(GL_UNPACK_ALIGNMENT, oldAlign);
        gl.glPixelStorei(GL_UNPACK_ALIGNMENT, 1);

        for (int mipmapLevel = 0; mipmapLevel < 8; mipmapLevel++) {

            int width = 128 >> mipmapLevel;
            int height = 128 >> mipmapLevel;

            byte[] currColor = mipmapColors[mipmapLevel];
            ByteBuffer buffer = fillWithColors(currColor, width, height);

            gl.glTexImage2D(GL_TEXTURE_2D, mipmapLevel, GL_RGB8, width, height, 0, GL_RGB, GL_UNSIGNED_BYTE, buffer);

            destroyBuffer(buffer);
        }

        gl.glPixelStorei(GL_UNPACK_ALIGNMENT, oldAlign.get(0));

        gl.glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_BASE_LEVEL, 0);
        gl.glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAX_LEVEL, 7);

        gl.glBindTexture(GL_TEXTURE_2D, 0);

        destroyBuffer(oldAlign);
    }

    private byte[][] mipmapColors = {
            {(byte) 0xFF, (byte) 0xFF, (byte) 0x00},
            {(byte) 0xFF, (byte) 0x00, (byte) 0xFF},
            {(byte) 0x00, (byte) 0xFF, (byte) 0xFF},
            {(byte) 0xFF, (byte) 0x00, (byte) 0x00},
            {(byte) 0x00, (byte) 0xFF, (byte) 0x00},
            {(byte) 0x00, (byte) 0x00, (byte) 0xFF},
            {(byte) 0x00, (byte) 0x00, (byte) 0x00},
            {(byte) 0xFF, (byte) 0xFF, (byte) 0xFF}};

    private ByteBuffer fillWithColors(byte[] color, int width, int height) {

        int numTexels = width * height;
        ByteBuffer buffer = GLBuffers.newDirectByteBuffer(numTexels * 3);

        byte red = color[0];
        byte green = color[1];
        byte blue = color[2];

        while(buffer.hasRemaining())
            buffer
                    .put(red)
                    .put(green)
                    .put(blue);

        buffer.position(0);
        return buffer;
    }

    private void createSamplers(GL3 gl) {

        gl.glGenSamplers(Sampler.MAX, samplerName);

        for (int samplerIx = 0; samplerIx < Sampler.MAX; samplerIx++) {

            gl.glSamplerParameteri(samplerName.get(samplerIx), GL_TEXTURE_WRAP_S, GL_REPEAT);
            gl.glSamplerParameteri(samplerName.get(samplerIx), GL_TEXTURE_WRAP_T, GL_REPEAT);
        }

        gl.glSamplerParameteri(samplerName.get(Sampler.Nearest), GL_TEXTURE_MAG_FILTER, GL_NEAREST);
        gl.glSamplerParameteri(samplerName.get(Sampler.Nearest), GL_TEXTURE_MIN_FILTER, GL_NEAREST);

        gl.glSamplerParameteri(samplerName.get(Sampler.Linear), GL_TEXTURE_MAG_FILTER, GL_LINEAR);
        gl.glSamplerParameteri(samplerName.get(Sampler.Linear), GL_TEXTURE_MIN_FILTER, GL_LINEAR);

        gl.glSamplerParameteri(samplerName.get(Sampler.Linear_MipMap_Nearest), GL_TEXTURE_MAG_FILTER, GL_LINEAR);
        gl.glSamplerParameteri(samplerName.get(Sampler.Linear_MipMap_Nearest), GL_TEXTURE_MIN_FILTER, GL_LINEAR_MIPMAP_NEAREST);

        gl.glSamplerParameteri(samplerName.get(Sampler.Linear_MipMap_Linear), GL_TEXTURE_MAG_FILTER, GL_LINEAR);
        gl.glSamplerParameteri(samplerName.get(Sampler.Linear_MipMap_Linear), GL_TEXTURE_MIN_FILTER, GL_LINEAR_MIPMAP_LINEAR);

        gl.glSamplerParameteri(samplerName.get(Sampler.LowAnysotropic), GL_TEXTURE_MAG_FILTER, GL_LINEAR);
        gl.glSamplerParameteri(samplerName.get(Sampler.LowAnysotropic), GL_TEXTURE_MIN_FILTER, GL_LINEAR_MIPMAP_LINEAR);
        gl.glSamplerParameterf(samplerName.get(Sampler.LowAnysotropic), GL_TEXTURE_MAX_ANISOTROPY_EXT, 4.0f);


        int maxAniso = caps.limits.MAX_TEXTURE_MAX_ANISOTROPY_EXT;

        System.out.println("Maximum anisotropy: " + maxAniso);

        gl.glSamplerParameteri(samplerName.get(Sampler.MaxAnysotropic), GL_TEXTURE_MAG_FILTER, GL_LINEAR);
        gl.glSamplerParameteri(samplerName.get(Sampler.MaxAnysotropic), GL_TEXTURE_MIN_FILTER, GL_LINEAR_MIPMAP_LINEAR);
        gl.glSamplerParameteri(samplerName.get(Sampler.MaxAnysotropic), GL_TEXTURE_MAX_ANISOTROPY_EXT, maxAniso);
    }

    @Override
    public void display(GL3 gl) {

        gl.glClearBufferfv(GL_COLOR, 0, clearColor.put(0, 0.75f).put(1, 0.75f).put(2, 1.0f).put(3, 1.0f));
        gl.glClearBufferfv(GL_DEPTH, 0, clearDepth.put(0, 1.0f));


        camTimer.update();

        float cyclicAngle = camTimer.getAlpha() * 6.28f;
        float hOffset = glm.cos(cyclicAngle) * .25f;
        float vOffset = glm.sin(cyclicAngle) * .25f;

        MatrixStack modelMatrix = new MatrixStack();

        Mat4 worldToCamMat = glm.lookAt(
                new Vec3(hOffset, 1f, -64f),
                new Vec3(hOffset, -5f + vOffset, -44f),
                new Vec3(0f, 1f, 0f));

        modelMatrix
                .applyMatrix(worldToCamMat)
                .push();
        {
            gl.glUseProgram(program.theProgram);

            gl.glUniformMatrix4fv(program.modelToCameraMatrixUL, 1, false, modelMatrix.top().to(matBuffer));

            gl.glActiveTexture(GL_TEXTURE0 + Semantic.Sampler.DIFFUSE);
            gl.glBindTexture(GL_TEXTURE_2D, textureName.get(useMipmapTexture ? Texture.MipmapTest : Texture.Checker));
            gl.glBindSampler(Semantic.Sampler.DIFFUSE, samplerName.get(currSampler));

            if (drawCorridor)
                corridor.render(gl, "tex");
            else
                plane.render(gl, "tex");

            gl.glBindSampler(Semantic.Sampler.DIFFUSE, 0);
            gl.glBindTexture(GL_TEXTURE_2D, 0);

            gl.glUseProgram(0);
        }
        modelMatrix.pop();
    }

    @Override
    public void reshape(GL3 gl, int w, int h) {

        MatrixStack persMatrix = new MatrixStack();
        persMatrix.perspective(90f, w / (float) h, 1f, 1000f);

        gl.glBindBuffer(GL_UNIFORM_BUFFER, projBufferName.get(0));
        gl.glBufferSubData(GL_UNIFORM_BUFFER, 0, Mat4.SIZE, persMatrix.top().to(matBuffer));
        gl.glBindBuffer(GL_UNIFORM_BUFFER, 0);

        gl.glViewport(0, 0, w, h);
    }

    @Override
    public void end(GL3 gl) {

        plane.dispose(gl);
        corridor.dispose(gl);

        gl.glDeleteProgram(program.theProgram);

        gl.glDeleteBuffers(1, projBufferName);
        gl.glDeleteTextures(Texture.MAX, textureName);
        gl.glDeleteSamplers(Sampler.MAX, samplerName);

        destroyBuffers(projBufferName, textureName, samplerName);
    }

    @Override
    public void keyPressed(KeyEvent ke) {

        switch (ke.getKeyCode()) {

            case KeyEvent.VK_ESCAPE:
                quit();
                break;

            case KeyEvent.VK_SPACE:
                useMipmapTexture = !useMipmapTexture;
                break;

            case KeyEvent.VK_Y:
                drawCorridor = !drawCorridor;
                break;

            case KeyEvent.VK_P:
                camTimer.togglePause();
                break;
        }

        if(KeyEvent.VK_1 <= ke.getKeyCode() && ke.getKeyCode() <= KeyEvent.VK_9) {
            int number = ke.getKeyCode() - KeyEvent.VK_1;
            if(number < Sampler.MAX) {
                System.out.println("Sampler: "+samplerNames[number]);
                currSampler = number;
            }
        }
    }

    interface Sampler {
        int Nearest = 0;
        int Linear = 1;
        int Linear_MipMap_Nearest = 2;
        int Linear_MipMap_Linear = 3;
        int LowAnysotropic = 4;
        int MaxAnysotropic = 5;
        int MAX = 6;
    }

    private String[] samplerNames = {"Nearest", "Linear", "Linear with nearest mipmaps", "Linear with linear mipmaps",
            "Low anisotropic", "Max anisotropic"};

    private class ProgramData {

        public int theProgram;
        public int modelToCameraMatrixUL;

        public ProgramData(GL3 gl, String vertex, String fragment) {

            theProgram = programOf(gl, getClass(), "tut15", vertex, fragment);

            modelToCameraMatrixUL = gl.glGetUniformLocation(theProgram, "modelToCameraMatrix");

            gl.glUniformBlockBinding(
                    theProgram,
                    gl.glGetUniformBlockIndex(theProgram, "Projection"),
                    Semantic.Uniform.PROJECTION);

            gl.glUseProgram(theProgram);
            gl.glUniform1i(
                    gl.glGetUniformLocation(theProgram, "colorTexture"),
                    Semantic.Sampler.DIFFUSE);
            gl.glUseProgram(0);
        }
    }
}

File: src/main/kotlin/main/tut15/manyImages.kt
package main.tut15

import com.jogamp.newt.event.KeyEvent
import com.jogamp.opengl.GL2ES3.*
import com.jogamp.opengl.GL2GL3.GL_UNSIGNED_INT_8_8_8_8_REV
import com.jogamp.opengl.GL3
import com.jogamp.opengl.GL3.GL_DEPTH_CLAMP
import com.jogamp.opengl.util.texture.spi.DDSImage
import glm.*
import glm.mat.Mat4
import glm.vec._3.Vec3
import main.framework.Framework
import main.framework.Semantic
import main.framework.component.Mesh
import uno.buffer.*
import uno.gl.getFloat
import uno.glm.MatrixStack
import uno.glsl.programOf
import uno.time.Timer
import java.io.File
import java.nio.ByteBuffer

/**
 * Created by GBarbieri on 31.03.2017.
 */

fun main(args: Array<String>) {
    ManyImages_().setup("Tutorial 15 - Many Images")
}

class ManyImages_ : Framework() {

    lateinit var program: ProgramData

    lateinit var plane: Mesh
    lateinit var corridor: Mesh

    val projBufferName = intBufferBig(1)

    object Texture {
        val Checker = 0
        val MipmapTest = 1
        val MAX = 2
    }

    val textureName = intBufferBig(Texture.MAX)
    val samplerName = intBufferBig(Sampler.MAX)

    val camTimer = Timer(Timer.Type.Loop, 5f)
    var useMipmapTexture = false
    var currSampler = 0
    var drawCorridor = false

    override fun init(gl: GL3) = with(gl) {

        initializeProgram(gl)

        plane = Mesh(gl, javaClass, "tut15/BigPlane.xml")
        corridor = Mesh(gl, javaClass, "tut15/Corridor.xml")

        glEnable(GL_CULL_FACE)
        glCullFace(GL_BACK)
        glFrontFace(GL_CW)

        val depthZNear = 0f
        val depthZFar = 1f

        glEnable(GL_DEPTH_TEST)
        glDepthMask(true)
        glDepthFunc(GL_LEQUAL)
        glDepthRangef(depthZNear, depthZFar)
        glEnable(GL_DEPTH_CLAMP)

        //Setup our Uniform Buffers
        glGenBuffers(1, projBufferName)
        glBindBuffer(GL_UNIFORM_BUFFER, projBufferName[0])
        glBufferData(GL_UNIFORM_BUFFER, Mat4.SIZE.L, null, GL_DYNAMIC_DRAW)

        glBindBufferRange(GL_UNIFORM_BUFFER, Semantic.Uniform.PROJECTION, projBufferName[0], 0, Mat4.SIZE.L)

        glBindBuffer(GL_UNIFORM_BUFFER, 0)

        // Generate all the texture names
        glGenTextures(Texture.MAX, textureName)

        loadCheckerTexture(gl)
        loadMipmapTexture(gl)
        createSamplers(gl)
    }

    fun initializeProgram(gl: GL3) {
        program = ProgramData(gl, "pt.vert", "tex.frag")
    }

    fun loadCheckerTexture(gl: GL3) = with(gl) {

        val file = File(javaClass.getResource("/tut15/checker.dds").toURI())

        val ddsImage = DDSImage.read(file)

        glBindTexture(GL_TEXTURE_2D, textureName[Texture.Checker])

        repeat(ddsImage.numMipMaps) { mipmapLevel ->

            val mipmap = ddsImage.getMipMap(mipmapLevel)

            glTexImage2D(GL_TEXTURE_2D, mipmapLevel, GL_RGB8, mipmap.width, mipmap.height, 0,
                    GL_BGRA, GL_UNSIGNED_INT_8_8_8_8_REV, mipmap.data)
        }

        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_BASE_LEVEL, 0)
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAX_LEVEL, ddsImage.numMipMaps - 1)
        glBindTexture(GL_TEXTURE_2D, 0)
    }

    fun loadMipmapTexture(gl: GL3) = with(gl) {

        glBindTexture(GL_TEXTURE_2D, textureName[Texture.MipmapTest])

        val oldAlign = intBufferBig(1)

        glGetIntegerv(GL_UNPACK_ALIGNMENT, oldAlign)
        glPixelStorei(GL_UNPACK_ALIGNMENT, 1)

        for (mipmapLevel in 0..7) {

            val width = 128 shr mipmapLevel
            val height = 128 shr mipmapLevel

            val currColor = mipmapColors[mipmapLevel]
            val buffer = fillWithColors(currColor, width, height)

            glTexImage2D(GL_TEXTURE_2D, mipmapLevel, GL_RGB8, width, height, 0, GL_RGB, GL_UNSIGNED_BYTE, buffer)

            buffer.destroy()
        }

        glPixelStorei(GL_UNPACK_ALIGNMENT, oldAlign[0])

        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_BASE_LEVEL, 0)
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAX_LEVEL, 7)

        glBindTexture(GL_TEXTURE_2D, 0)

        oldAlign.destroy()
    }

    val mipmapColors = arrayOf(
            byteArrayOf(0xFF.b, 0xFF.b, 0x00.b),
            byteArrayOf(0xFF.b, 0x00.b, 0xFF.b),
            byteArrayOf(0x00.b, 0xFF.b, 0xFF.b),
            byteArrayOf(0xFF.b, 0x00.b, 0x00.b),
            byteArrayOf(0x00.b, 0xFF.b, 0x00.b),
            byteArrayOf(0x00.b, 0x00.b, 0xFF.b),
            byteArrayOf(0x00.b, 0x00.b, 0x00.b),
            byteArrayOf(0xFF.b, 0xFF.b, 0xFF.b))

    fun fillWithColors(color: ByteArray, width: Int, height: Int): ByteBuffer {

        val numTexels = width * height
        val buffer = byteBufferBig(numTexels * 3)

        val (red, green, blue) = color

        while (buffer.hasRemaining())
            buffer
                    .put(red)
                    .put(green)
                    .put(blue)

        buffer.position(0)
        return buffer
    }

    fun createSamplers(gl: GL3) = with(gl) {

        glGenSamplers(Sampler.MAX, samplerName)

        repeat(Sampler.MAX) {

            glSamplerParameteri(samplerName[it], GL_TEXTURE_WRAP_S, GL_REPEAT)
            glSamplerParameteri(samplerName[it], GL_TEXTURE_WRAP_T, GL_REPEAT)
        }

        glSamplerParameteri(samplerName[Sampler.Nearest], GL_TEXTURE_MAG_FILTER, GL_NEAREST)
        glSamplerParameteri(samplerName[Sampler.Nearest], GL_TEXTURE_MIN_FILTER, GL_NEAREST)

        glSamplerParameteri(samplerName[Sampler.Linear], GL_TEXTURE_MAG_FILTER, GL_LINEAR)
        glSamplerParameteri(samplerName[Sampler.Linear], GL_TEXTURE_MIN_FILTER, GL_LINEAR)

        glSamplerParameteri(samplerName[Sampler.Linear_MipMap_Nearest], GL_TEXTURE_MAG_FILTER, GL_LINEAR)
        glSamplerParameteri(samplerName[Sampler.Linear_MipMap_Nearest], GL_TEXTURE_MIN_FILTER, GL_LINEAR_MIPMAP_NEAREST)

        glSamplerParameteri(samplerName[Sampler.Linear_MipMap_Linear], GL_TEXTURE_MAG_FILTER, GL_LINEAR)
        glSamplerParameteri(samplerName[Sampler.Linear_MipMap_Linear], GL_TEXTURE_MIN_FILTER, GL_LINEAR_MIPMAP_LINEAR)

        glSamplerParameteri(samplerName[Sampler.LowAnysotropic], GL_TEXTURE_MAG_FILTER, GL_LINEAR)
        glSamplerParameteri(samplerName[Sampler.LowAnysotropic], GL_TEXTURE_MIN_FILTER, GL_LINEAR_MIPMAP_LINEAR)
        glSamplerParameterf(samplerName[Sampler.LowAnysotropic], GL_TEXTURE_MAX_ANISOTROPY_EXT, 4.0f)


        val maxAniso = getFloat(GL_MAX_TEXTURE_MAX_ANISOTROPY_EXT)

        println("Maximum anisotropy: " + maxAniso)

        glSamplerParameteri(samplerName[Sampler.MaxAnysotropic], GL_TEXTURE_MAG_FILTER, GL_LINEAR)
        glSamplerParameteri(samplerName[Sampler.MaxAnysotropic], GL_TEXTURE_MIN_FILTER, GL_LINEAR_MIPMAP_LINEAR)
        glSamplerParameteri(samplerName[Sampler.MaxAnysotropic], GL_TEXTURE_MAX_ANISOTROPY_EXT, maxAniso.i)
    }

    override fun display(gl: GL3) = with(gl) {

        glClearBufferfv(GL_COLOR, 0, clearColor.put(0.75f, 0.75f, 1.0f, 1.0f))
        glClearBufferfv(GL_DEPTH, 0, clearDepth.put(0, 1.0f))


        camTimer.update()

        val cyclicAngle = camTimer.getAlpha() * 6.28f
        val hOffset = glm.cos(cyclicAngle) * .25f
        val vOffset = glm.sin(cyclicAngle) * .25f

        val modelMatrix = MatrixStack()

        val worldToCamMat = glm.lookAt(
                Vec3(hOffset, 1f, -64f),
                Vec3(hOffset, -5f + vOffset, -44f),
                Vec3(0f, 1f, 0f))

        modelMatrix.applyMatrix(worldToCamMat) run {

            glUseProgram(program.theProgram)

            glUniformMatrix4fv(program.modelToCameraMatrixUL, 1, false, top() to matBuffer)

            glActiveTexture(GL_TEXTURE0 + Semantic.Sampler.DIFFUSE)
            glBindTexture(GL_TEXTURE_2D, textureName[if (useMipmapTexture) Texture.MipmapTest else Texture.Checker])
            glBindSampler(Semantic.Sampler.DIFFUSE, samplerName[currSampler])

            if (drawCorridor)
                corridor.render(gl, "tex")
            else
                plane.render(gl, "tex")

            glBindSampler(Semantic.Sampler.DIFFUSE, 0)
            glBindTexture(GL_TEXTURE_2D, 0)

            glUseProgram(program.theProgram)
        }
    }

    override fun reshape(gl: GL3, w: Int, h: Int) = with(gl) {

        val persMatrix = MatrixStack()
        persMatrix.perspective(90f, w / h.f, 1f, 1000f)

        glBindBuffer(GL_UNIFORM_BUFFER, projBufferName[0])
        glBufferSubData(GL_UNIFORM_BUFFER, 0, Mat4.SIZE.L, persMatrix.top() to matBuffer)
        glBindBuffer(GL_UNIFORM_BUFFER, 0)

        glViewport(0, 0, w, h)
    }

    override fun end(gl: GL3) = with(gl) {

        plane.dispose(gl)
        corridor.dispose(gl)

        glDeleteProgram(program.theProgram)

        glDeleteBuffers(1, projBufferName)
        glDeleteTextures(Texture.MAX, textureName)
        glDeleteSamplers(Sampler.MAX, samplerName)

        destroyBuffers(projBufferName, textureName, samplerName)
    }

    override fun keyPressed(ke: KeyEvent) {

        when (ke.keyCode) {

            KeyEvent.VK_ESCAPE -> quit()

            KeyEvent.VK_SPACE -> useMipmapTexture = !useMipmapTexture

            KeyEvent.VK_Y -> drawCorridor = !drawCorridor

            KeyEvent.VK_P -> camTimer.togglePause()
        }

        if (ke.keyCode in KeyEvent.VK_1 .. KeyEvent.VK_9) {
            val number = ke.keyCode - KeyEvent.VK_1
            if (number < Sampler.MAX) {
                println("Sampler: " + samplerNames[number])
                currSampler = number
            }
        }
    }

    object Sampler {
        val Nearest = 0
        val Linear = 1
        val Linear_MipMap_Nearest = 2
        val Linear_MipMap_Linear = 3
        val LowAnysotropic = 4
        val MaxAnysotropic = 5
        val MAX = 6
    }

    val samplerNames = arrayOf("Nearest", "Linear", "Linear with nearest mipmaps", "Linear with linear mipmaps", "Low anisotropic", "Max anisotropic")

    class ProgramData(gl: GL3, vertex: String, fragment: String) {

        val theProgram = programOf(gl, javaClass, "tut15", vertex, fragment)
        val modelToCameraMatrixUL = gl.glGetUniformLocation(theProgram, "modelToCameraMatrix")

        init {
            with(gl) {
                glUniformBlockBinding(
                        theProgram,
                        glGetUniformBlockIndex(theProgram, "Projection"),
                        Semantic.Uniform.PROJECTION)

                glUseProgram(theProgram)
                glUniform1i(
                        glGetUniformLocation(theProgram, "colorTexture"),
                        Semantic.Sampler.DIFFUSE)
                glUseProgram(0)
            }
        }
    }
}